IMPORT THE REQUIRED LIBERARIES

In [27]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

LOAD THE DATASET 

In [28]:
dataset = pd.read_csv("Old_Car.csv")

PRINT DATASET 

In [29]:
dataset.head(2)

Unnamed: 0                              Name Location  Year  \
0           0            Maruti Wagon R LXI CNG   Mumbai  2010   
1           1  Hyundai Creta 1.6 CRDi SX Option     Pune  2015   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              72000       CNG       Manual      First  26.6 km/kg   998 CC   
1              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   

       Power  Seats  Price  
0  58.16 bhp    5.0   1.75  
1  126.2 bhp    5.0  12.50

DELETE COLUMNS WHICH ARE NOT SO USEFUL

In [30]:
dataset.drop(columns = ["Unnamed: 0","Location","Kilometers_Driven","Mileage","Engine","Power","Seats"],inplace = True)

AGAIN CHECK DATASET

In [31]:
dataset.head(2)

Name  Year Fuel_Type Transmission Owner_Type  \
0            Maruti Wagon R LXI CNG  2010       CNG       Manual      First   
1  Hyundai Creta 1.6 CRDi SX Option  2015    Diesel       Manual      First   

   Price  
0   1.75  
1  12.50

NOW EXTRACT INFORMATION REGARDING THE DATASET

In [32]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6019 entries, 0 to 6018
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          6019 non-null   object 
 1   Year          6019 non-null   int64  
 2   Fuel_Type     6019 non-null   object 
 3   Transmission  6019 non-null   object 
 4   Owner_Type    6019 non-null   object 
 5   Price         6019 non-null   float64
dtypes: float64(1), int64(1), object(4)
memory usage: 282.3+ KB


NOW CHECK MISSING VALUES IF ANY

In [33]:
dataset.isnull().sum()

Name            0
Year            0
Fuel_Type       0
Transmission    0
Owner_Type      0
Price           0
dtype: int64

NOW FIND THE STATISTICS PARAMETERS REGARDING DATASET

In [34]:
dataset.describe()

Year        Price
count  6019.000000  6019.000000
mean   2013.358199     9.479468
std       3.269742    11.187917
min    1998.000000     0.440000
25%    2011.000000     3.500000
50%    2014.000000     5.640000
75%    2016.000000     9.950000
max    2019.000000   160.000000

PERFORM ENCODING ON DATASET(convert categorical data into numeric form)
BY USING LABEL ENCODER.


In [35]:
from sklearn.preprocessing import LabelEncoder

In [36]:
Car_Name_le  = LabelEncoder()

In [37]:
dataset["Name"]  = Car_Name_le.fit_transform(dataset["Name"])

In [38]:
dataset["Owner_Type"].unique()

array(['First', 'Second', 'Fourth & Above', 'Third'], dtype=object)

In [39]:
Owner_type_le = LabelEncoder()

In [40]:
dataset["Owner_Type"] = Owner_type_le.fit_transform(dataset["Owner_Type"])

In [41]:
dataset["Fuel_Type"].unique()

array(['CNG', 'Diesel', 'Petrol', 'LPG', 'Electric'], dtype=object)

In [42]:
Fuel_Type_le = LabelEncoder()

In [43]:
dataset["Fuel_Type"] = Fuel_Type_le.fit_transform(dataset["Fuel_Type"])

In [44]:
Transmission_le = LabelEncoder()

In [45]:
dataset["Transmission"] = Transmission_le.fit_transform(dataset["Transmission"])

NOW PRINT THE NEW DATASET WHICH WE HAVE ENCODED

In [46]:
dataset.head(3)

Name  Year  Fuel_Type  Transmission  Owner_Type  Price
0  1200  2010          0             1           0   1.75
1   512  2015          1             1           0  12.50
2   486  2011          4             1           0   4.50

CONVERT THE DATASET INTO INPUT AND OUTPUT DATA

In [47]:
Input_data = dataset.iloc[:,:-1]
OutPut_data = dataset["Price"]


FEATURE SCALING

In [48]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()

PERFORM FIT AND TRANSFORM AND  THEN OVERWRITE IN Input_data VARIABLE

In [49]:
Input_data = pd.DataFrame(ss.fit_transform(Input_data),columns = Input_data.columns)

NOW SPLIT THE DATASET FOR TRAINING AND TESTING

In [50]:
from sklearn.model_selection import train_test_split

In [51]:
x_train,x_test,y_train,y_test = train_test_split(Input_data,OutPut_data,test_size = 0.2,random_state = 42)

IMPORT THE MACHINE LEARNING MODEL 

In [54]:
from sklearn.linear_model import LinearRegression,Lasso,Ridge,ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor

FIT THE MODEL AND TEST THE ACCURACY OF THE MODEL(FOR ALL ABOVE MODELS)

In [55]:
lr = LinearRegression()
lr.fit(x_train,y_train)
lr.score(x_train,y_train)*100 , lr.score(x_test,y_test)*100

(45.367331272054834, 43.36202734659202)

In [56]:
la = Lasso(alpha = 0.5)
la.fit(x_train,y_train)
la.score(x_train,y_train)*100,la.score(x_test,y_test)*100

(44.69961990605347, 42.91587573650906)

In [57]:
ri = Ridge(alpha = 0.5)
ri.fit(x_train,y_train)
ri.score(x_train,y_train)*100 , ri.score(x_test,y_test)*100

(45.36733090213709, 43.362353928520946)

In [58]:
el = ElasticNet(alpha = 0.5)
el.fit(x_train,y_train)
el.score(x_train,y_train)*100 ,el.score(x_test,y_test)*100

(43.2368049283202, 41.85994148455223)

In [61]:
dt = DecisionTreeRegressor()
dt.fit(x_train,y_train)
dt.score(x_train,y_train)*100 ,dt.score(x_test,y_test)*100

(99.54790090824429, 63.21415482999666)

In [62]:
rf = RandomForestRegressor(n_estimators = 40)
rf.fit(x_train,y_train)
rf.score(x_train,y_train)*100 ,rf.score(x_test,y_test)*100

(97.09608762703284, 80.74389280349766)

In [63]:
sv = SVR()
sv.fit(x_train,y_train)
sv.score(x_train,y_train)*100 ,sv.score(x_test,y_test)*100

(53.124650564523044, 52.612208143601634)

In [64]:
kn = KNeighborsRegressor(n_neighbors = 5)
kn.fit(x_train,y_train)
kn.score(x_train,y_train)*100 ,kn.score(x_test,y_test)*100

(77.09022630069194, 68.09409083341242)

CALCULATE ERROR FOR THE HIGHEST ACCURACY MODEL

In [65]:
from sklearn.metrics import mean_squared_error,mean_absolute_error

In [66]:
mean_squared_error(y_test,dt.predict(x_test)),mean_absolute_error(y_test,dt.predict(x_test))

(45.26847415483729, 2.279401279045797)

In [67]:
mean_squared_error(y_test,rf.predict(x_test)),mean_absolute_error(y_test,rf.predict(x_test))

(23.696467674431947, 1.950105603859778)

NOW PREDICT THE PRICE FOR ANY PARTICULAR CAR

In [68]:
x_test

Name      Year  Fuel_Type  Transmission  Owner_Type
2868  0.413435 -0.109559  -0.905771      0.632529   -0.463672
5924 -0.098110  1.113881  -0.905771      0.632529   -0.463672
3764  0.015147  0.196301  -0.905771      0.632529   -0.463672
4144  1.272301  0.808021  -0.905771      0.632529    1.980149
2780 -0.619093 -1.332998   1.087831      0.632529   -0.463672
...        ...       ...        ...           ...         ...
5926  0.052899 -4.085738   1.087831      0.632529   -0.463672
4216  0.549343  0.502161   1.087831      0.632529   -0.463672
1351  0.215235  0.502161   1.087831      0.632529   -0.463672
4603  0.251099  1.725601  -1.570306      0.632529   -0.463672
5668 -1.194816 -1.944718  -0.905771      0.632529   -0.463672

[1204 rows x 5 columns]

In [69]:
y_test

2868     5.75
5924    10.08
3764     7.85
4144     2.40
2780     1.60
        ...  
5926     0.55
4216     4.65
1351     6.93
4603     5.38
5668     1.99
Name: Price, Length: 1204, dtype: float64

In [98]:
rf.predict([[0.413435 ,	-0.109559 ,	-0.905771 ,	0.632529 ,	-0.463672]])

C:\Users\HAMDAAN INFOTECH\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


array([4.9634127])

In [99]:
new_data1 = pd.DataFrame([["Maruti Wagon R LXI CNG", 	2010,"CNG" ,"Manual", 	"First"]],columns= x_train.columns)

In [100]:
new_data1

Name  Year Fuel_Type Transmission Owner_Type
0  Maruti Wagon R LXI CNG  2010       CNG       Manual      First

In [102]:
new_data1["Name"] = Car_Name_le.transform(new_data1["Name"])

In [103]:
new_data1["Fuel_Type"] = Fuel_Type_le.transform(new_data1["Fuel_Type"])

In [104]:
new_data1["Transmission"] = Transmission_le.transform(new_data1["Transmission"])

In [106]:
new_data1["Owner_Type"] = Owner_type_le.transform(new_data1["Owner_Type"])

In [108]:
new_data1 = pd.DataFrame(ss.transform(new_data1),columns = new_data1.columns)

In [109]:
rf.predict(new_data1)


array([1.825])

#array([1.825])  the predicted price for the "Maruti Wagon R LXI CNG" car
#which is approximately equal to the price of "Maruti Wagon R LXI CNG" 
#for which we want to calculate the price (i.e 1.75 in dataset)
